docs/source/notes/autograd.rst

```diff
@@ -815,3 +815,64 @@ Without the hooks, ``x``, ``y.grad_fn._saved_self`` and
 ``y.grad_fn._saved_other`` all refer to the same tensor object.
 With the hooks, PyTorch will pack and unpack `x` into two new tensor objects
 that share the same storage with the original `x` (no copy performed).
+
+.. _backward-hooks-execution:
+
+Backward Hooks execution
+------------------------
+
+This section will discuss when different hooks fire or don't fire.
+Then it will discuss the order in which they are fired.
+The hooks that will be covered are: Pre-hooks registered to Tensor via
+:meth:`torch.tensor.register_hook`,
+post-hooks registered to Node via :meth:`torch.autograd.graph.Node.register_hook`, and
+pre-hooks registered to Node via :meth:`torch.autograd.graph.Node.register_prehook`.
+
+Whether a particular hook will be fired
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Hooks registered to a Tensor via :meth:`torch.tensor.register_hook`
+are executed when gradients are being computed for that Tensor. This does not require
+the Node corresponding to that Tensor to be executed.
+
+Hooks registered to :class:`torch.autograd.graph.Node` using
+:meth:`torch.autograd.graph.Node.register_hook` or
+:meth:`torch.autograd.graph.Node.register_prehook` on the other hand, are only fired if
+the Node it was registered to is executed.
+
+Whether a particular Node is executed may depend on whether the backward pass was called with
+:func:`torch.autograd.grad` or :func:`torch.autograd.backward`.
+Specifically, you should be aware of these differences when you register a hook on a
+Node corresponding to a Tensor that you are passing to :func:`torch.autograd.grad` or
+:func:`torch.autograd.backward` as part of the ``inputs`` argument.
+
+If you are using :func:`torch.autograd.backward`, all of the above mentioned hooks will be executed,
+whether or not you specified the ``inputs`` argument. This is because `.backward()` executes all
+Nodes, even if they correspond to a Tensor specified as an input.
+(Note that the execution of this additional Node corresponding to Tensors passed as  ``inputs``
+is usually unnecessary, but done anyway. This behavior is subject to change,
+you should not depend on it.)
+
+On the other hand, if you are using :func:`torch.autograd.grad`, the backward hooks registered
+to Nodes that correspond to the Tensors passed to ``input`` may not be executed, because
```

Comment:
may not be executed: will not be executed due to this input (it might still be executed due to another input)